modules/rostests/apitests/user32/PrivateExtractIcons.c

```diff
@@ -3,41 +3,128 @@
+ *                  Doug Lyons <douglyons@douglyons.com>
  */
 
 #include "precomp.h"
+#include <stdio.h>
+
+BOOL FileExists(LPCSTR FileName)
+{
+    FILE *fp = NULL;
+    BOOL exists = FALSE;
+
+    fp = fopen(FileName, "r");
+    if (fp != NULL)
+    {
+        exists = TRUE;
+        fclose(fp);
+    }
+    return exists;
+}
+
+BOOL ResourceToFile(INT i, LPCSTR FileName)
+{
+    FILE *fout;
+    HGLOBAL hData;
+    HRSRC hRes;
+    LPVOID lpResLock;
+    UINT iSize;
+
+    if (FileExists(FileName))
+    {
+        skip("'%s' already exists. Exiting now\n", FileName);
+        return FALSE;
+    }
+
+    hRes = FindResourceW(NULL, MAKEINTRESOURCEW(i), MAKEINTRESOURCEW(RT_RCDATA));
+    if (hRes == NULL)
+    {
+        skip("Could not locate resource (%d). Exiting now\n", i);
+        return FALSE;
+    }
+
+    iSize = SizeofResource(NULL, hRes);
+
+    hData = LoadResource(NULL, hRes);
+    if (hData == NULL)
+    {
+        skip("Could not load resource (%d). Exiting now\n", i);
+        return FALSE;
+    }
+
+    // Lock the resource into global memory.
+    lpResLock = LockResource(hData);
+    if (lpResLock == NULL)
+    {
+        skip("Could not lock resource (%d). Exiting now\n", i);
+        return FALSE;
+    }
+
+    fout = fopen(FileName, "wb");
+    fwrite(lpResLock, iSize, 1, fout);
+    fclose(fout);
+    return TRUE;
+}
 
 static struct
 {
     PCWSTR FilePath;
     UINT cIcons;        // Return value from PrivateExtractIconsW
+    UINT cTotIcons;     // Return value of Total Icons in File
     BOOL bhIconValid;   // Whether or not the returned icon handle is not NULL.
 } IconTests[] =
 {
     /* Executables with icons */
-    {L"notepad.exe", 1, TRUE},
-    {L"%SystemRoot%\\System32\\cmd.exe", 1, TRUE},
+    {L"notepad.exe", 1, 1, TRUE},
+    {L"%SystemRoot%\\System32\\cmd.exe", 1, 1, TRUE},
 
     /* Executable without icon */
-    {L"%SystemRoot%\\System32\\autochk.exe", 0, FALSE},
+    {L"%SystemRoot%\\System32\\autochk.exe", 0, 0, FALSE},
 
-    /* Existing file */
-    {L"%SystemRoot%\\System32\\shell32.dll", 1, TRUE},
+    /* Existing file 233 is ROS Only */
+    {L"%SystemRoot%\\System32\\shell32.dll", 1, 233, TRUE},
 
     /* Non-existing files */
-    {L"%SystemRoot%\\non-existent-file.sdf", 0xFFFFFFFF, FALSE},
+    {L"%SystemRoot%\\non-existent-file.sdf", 0xFFFFFFFF, 0, FALSE},
+    /* Multiple icons in the same EXE file (18 icons) */
+    {L"%SystemRoot%\\explorer.exe", 1, 18, TRUE},
+
+    /* Multiple icons in the same ICO file (6 icons) */
+    {L"%SystemRoot%\\bin\\sysicon.ico", 1, 1, TRUE},
+
+    /* ICO file with both normal and PNG icons */
+    {L"%SystemRoot%\\bin\\ROS.ico", 1, 1, TRUE},
 };
 
 START_TEST(PrivateExtractIcons)
 {
     HICON ahIcon;
-    UINT i, aIconId, cIcons;
+    UINT i, aIconId, cIcons, cTotIcons;
+    CHAR FileName[2][13] = { "ROS.ico", "sysicon.ico" };
+
+    if (!ResourceToFile(1, FileName[0]))
+        return;
+    if (!ResourceToFile(2, FileName[1]))
```

Comment:
Add `DeleteFileA(FileName[0]);`.

Reply:
I chose to delete both files at the end of the main program function. They both need to be accessed from within the test loops, so deleting the file here would not work. See lines 148 and 149 for where they are both deleted. Thanks.

Reply:
?? How is line 148 executed after line 109 `return`?